lib/persist-state.c

```diff
@@ -267,6 +314,12 @@ _alloc_value(PersistState *self, guint32 orig_size, gboolean in_use, guint8 vers
   persist_state_unmap_entry(self, self->current_ofs);
 
   self->current_ofs += size + sizeof(PersistValueHeader);
+
+  if (!_check_watermark(self) && !_grow_store(self, self->current_size + PERSIST_FILE_INITIAL_SIZE))
+    {
+      main_loop_exit();
+    }
+
   return result;
```

Comment:
well, what I see there is a concrete function call, one that creates a
static dependency between persiststate and the mainloop.

the fact that mainloop in turn uses an ivykis event is an implementation
detail of mainloop.

but the goal here is not to argue about aesthetics, but rather whether you
could test persist state with unit tests without having to pull in mainloop
or the rest of the core. earlier this was possible, with this change it
wouldn't be unless someone mocked out main_loop_exit() in the unit test.

that's my argument. with all this said I could probably live with another
circular dependency, but at least let's agree that it is not nice.
Cheers,
Bazsi

-- 
Bazsi

On Mon, Nov 30, 2015 at 11:30 AM, juhaszviktor <notifications@github.com>
wrote:

> In lib/persist-state.c
> <https://github.com/balabit/syslog-ng/pull/717#discussion_r46127314>:
>
> > @@ -267,6 +314,12 @@ _alloc_value(PersistState *self, guint32 orig_size, gboolean in_use, guint8 vers
> >    persist_state_unmap_entry(self, self->current_ofs);
> >
> >    self->current_ofs += size + sizeof(PersistValueHeader);
> > +
> > +  if (!_check_watermark(self) && !_grow_store(self, self->current_size + PERSIST_FILE_INITIAL_SIZE))
> > +    {
> > +      main_loop_exit();
> > +    }
> > +
> >    return result;
>
> @bazsi <https://github.com/bazsi>, sorry, but I think that main_loop_exit
> is a callback, because this is an ivykis event, where mainloop registers,
> do I misunderstand something?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/717/files#r46127314>.
>